torch/csrc/autograd/profiler.cpp

```diff
@@ -13,135 +13,260 @@
+        cuda_stubs->nvtxRangePushA(s.str().c_str());
+      } else {
+        cuda_stubs->nvtxRangePushA(name.str());
       }
-      cuda_stubs->nvtxRangePushA(s.str().c_str());
     } else {
-      cuda_stubs->nvtxRangePushA(name.str());
+      std::lock_guard<std::mutex> guard(mutex_);
+      auto& list = getEventList();
+      list.record(
+          EventKind::PushRange,
+          name,
+          RecordFunction::currentThreadId(),
+          config_.state == ProfilerState::CUDA,
+          std::move(shapes));
     }
-  } else {
-    getEventList().record(
-        EventKind::PushRange,
-        name,
-        thread_id,
-        state == ProfilerState::CUDA,
-        std::move(shapes));
   }
-}
 
-void popRange() {
-  if (state == ProfilerState::Disabled) {
-    return;
-  }
-  if (state == ProfilerState::NVTX) {
-    cuda_stubs->nvtxRangePop();
-  } else {
-    getEventList().record(
-        EventKind::PopRange,
-        StringView(""),
-        thread_id,
-        state == ProfilerState::CUDA);
+  void popRange(uint64_t orig_thread_id) {
+    if (config_.state == ProfilerState::Disabled) {
+      return;
+    }
+    if (config_.state == ProfilerState::NVTX) {
+      cuda_stubs->nvtxRangePop();
+    } else {
+      std::lock_guard<std::mutex> guard(mutex_);
+      auto& list = getEventList(orig_thread_id);
+      list.record(
+          EventKind::PopRange,
+          StringView(""),
+          orig_thread_id,
+          config_.state == ProfilerState::CUDA);
+    }
   }
-}
 
-void enableProfiler(ProfilerConfig config) {
-  ProfilerState new_state = config.state;
-  AT_ASSERT(new_state != ProfilerState::Disabled);
-  if (new_state == ProfilerState::NVTX && !cuda_stubs->enabled())
-    throw std::runtime_error("Can't use NVTX profiler - PyTorch was compiled without CUDA");
-  if (state != ProfilerState::Disabled && new_state != state) {
-    throw std::runtime_error("can't change kind of profiling (e.g. NVTX to CPU) while profiler is running");
+ private:
+  // not thread safe
+  RangeEventList& getEventList(int64_t thread_id = -1) {
+    if (thread_id < 0) {
+      thread_id = RecordFunction::currentThreadId();
+    }
+    auto it = event_lists_map_.find(thread_id);
+    if (it != event_lists_map_.end()) {
+      return *(it->second);
+    } else {
+      auto event_list = std::make_shared<RangeEventList>();
+      event_lists_map_[thread_id] = event_list;
+      return *(event_list.get());
+    }
   }
 
+  std::mutex mutex_;
+  std::unordered_map<uint16_t, std::shared_ptr<RangeEventList>>
+      event_lists_map_;
+  ProfilerConfig config_ = ProfilerConfig(ProfilerState::Disabled, false);
+};
+
+void pushProfilingCallbacks(bool needs_inputs = false) {
   pushCallback(
-      [config](const RecordFunction& fn) {
+      [needs_inputs](const RecordFunction& fn) {
+        const auto& state = at::ThreadLocalDebugInfo::get(at::DebugInfoKind::PROFILER_STATE);
+        auto state_ptr = dynamic_cast<ProfilerThreadLocalState*>(state.get());
```

Comment:
maybe we can do static_cast here? It's a bit faster and should be "safe"

Reply:
yep, will do